improve the parser and tests

fwyzard opened the pull request:
TriggerExpression parsers:
  - improve the representation of trigger expression
  - delimit tokens and operands

Unit test:
  - print how the expressions are interpreted
  - compare the parsed representation with the expected one
  - add more tests

fwyzard committed:
Print how the expressions are interpreted

fwyzard committed:
Improve the representation of trigger expression

Improved the dumped representation of trigger expressions:
  - AND, OR, and NOT operators add parentheses around their
    subexpression;
  - uninitialised Path and L1uGT readers use an explicit name
    instead of FALSE.

fwyzard committed:
Delimit tokens and operands

Require a delimiter after each token and operamd, to avoid en expression
like:

    NOTSomePath ANDOtherPath

to be parsed as:

    NOT SomePath AND OtherPath

fwyzard committed:
Add more tests

fwyzard committed:
Compare the parsed and expected representations

missirol commented on HLTrigger/HLTcore/interface/TriggerExpressionParser.h:
```diff
@@ -37,16 +48,24 @@ namespace triggerExpression {
       operand %= (prescale | element);
 
-      unary = ((qi::lit("NOT") >> operand)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
+      unary = ((operand_not >> operand)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
```
```suggestion
      unary = ((operand_not >> expression)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
```
Playing around, it seemed like this could fix the "NOT NOT [expr]" case.

fwyzard replied:
Ah... non sapevo si potesso usare rules ricorsive

fwyzard replied:
Se funziona, fai un altro commit (credo ci sia anche un test da cambiare) :+1:

missirol replied:
Okay, allora intanto faccio il merge di questa, poi ricontrollo il tutto.

missirol replied:
> Playing around, it seemed like this could fix the "NOT NOT [expr]" case.

https://github.com/missirol/cmssw/commit/17ffd443196392da9baec10ef779bfa246fe420a contiene questo update, ma ha un limite: cambia il parsing di `NOT THIS OR THAT` in `NOT (THIS OR THAT)`, mentre al momento sarebbe `(NOT THIS) OR THAT`. Forse e' meglio non cambiare quest'ultimo, che e' piu' intuitivo, solo per avere la possibilita' di scrivere `NOT NOT` invece di `NOT (NOT (`, dato che `NOT NOT` sara' raramente usato.

fwyzard replied:
No, infatti: direi di che l'interpretazione che ci si aspetta sia `(NOT THIS) OR THAT`.

missirol replied:
Okay, quindi `NOT THIS OR THAT` rimarra' `(NOT THIS) OR THAT`.

Ora sto provando
```diff
-      unary = ((operand_not >> operand)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
+      unary = ((operand_not >> unary)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
```
(che, pensandoci, ha piu' senso) e sembra funzionare correttamente.

fwyzard replied:
OK, questo mi sembra sensato.

fwyzard committed:
Support NOT NOT expression

Acceppt an expression with an arbitrary number of consecutive NOT
operands, and discard any even number of them.
Thus

    NOT NOT expression
    NOT NOT NOT expression

are parsed as

    expression
    NOT expression

and so on.

missirol commented:
Closing. All suggestions were included in https://github.com/cms-sw/cmssw/pull/39108. Thanks!
